### PR TITLE
Test interaction with ledger using sandbox and json api

### PR DESCRIPTION
**Pull request by rohanjr**
Not a fully end-to-end test as it doesn't use a browser. Instead this interacts with the sandbox through the daml/ledger library and the json api. I will extend the tests to use a headless browser with puppeteer next.

### CODE REVIEW

**Comment by rohanjr** [comment]
Hmm, the test procedure hangs in CI (after both tests pass).
```
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests.
```

File: ui/src/index.test.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+let jsonApiProc: ChildProcess | undefined = undefined;
+let startProc:   ChildProcess | undefined = undefined;
+
+beforeAll(async () => {
+});
+
+afterAll(() => {
+  // Shut down all possibly running daml processes
+  if (startProc) {
+    startProc.kill("SIGTERM");
```
Comment (cocreature) [inline ui/src/index.test.tsx]: Does this work on Windows?

File: ui/src/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+const JSON_API_PORT = 7575;
+
+let sandboxProc: ChildProcess | undefined = undefined;
+let jsonApiProc: ChildProcess | undefined = undefined;
+let startProc:   ChildProcess | undefined = undefined;
+
+beforeAll(async () => {
+});
+
+afterAll(() => {
```
Comment (cocreature) [inline ui/src/index.test.tsx]: Why are we killing the processes in `afterAll` but not starting them in `beforeAll`?

File: ui/src/index.test.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+});
+
+const fail = () => expect(false).toBeTruthy();
+
+// Extend timeout to allow sandbox and json api server to start running
+jest.setTimeout(20_000);
+
+it('starts sandbox and json api server', async () => {
+  // Set up enviroment to run daml assistant commands
+  const env = Object.assign(process.env, {PATH: process.env.HOME + '/.daml/bin:' + process.env.PATH});
```
Comment (cocreature) [inline ui/src/index.test.tsx]: I don’t understand this. Why is this not run in an environment where this is already in PATH?

Comment (cocreature) [inline ui/src/index.test.tsx]: Also looks fishy on Windows :slightly_smiling_face:

File: ui/src/index.test.tsx (hunk 2 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: Starting the processes will move to `beforeAll`, did it this way while I only had one test.

File: ui/src/index.test.tsx (hunk 3 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: Hmm I suppose the path should already be set this way. I need to set the path here for the test to pass within VSCode - I guess it doesn't pick up the path there.

File: ui/src/index.test.tsx (hunk 1 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: I don't know (so probably not :P). I'll test it once the other issues are fixed, esp the processes not getting shut down in CI (Linux).

**Comment by rohanjr** [comment]
With @hurryabit 's blessing, I'll ignore Windows for now and get the end-to-end tests running on Unixey systems for now. The CI server is Ubuntu from what I see so we'll use this as the test for now.

**Comment by rohanjr** [comment]
Using `spawn` instead of `exec` and the `stdio: 'inherit'` option seems to have avoided the `Jest did not exit` error so I'll run with this.

File: ui/src/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const DAR_PATH = '.daml/dist/create-daml-app-0.1.0.dar';
+const LEDGER_ID = 'create-daml-app-sandbox';
+const APPLICATION_ID = 'create-daml-app';
+const SANDBOX_PORT = 6865;
+const JSON_API_PORT = 7575;
+
+let sandboxProc: ChildProcess | undefined = undefined;
+let jsonApiProc: ChildProcess | undefined = undefined;
+
+// Start a fresh sandbox and json api server for each test to have a clean slate
```
Comment (hurryabit) [inline ui/src/index.test.tsx]: What's the reason for _not_ using `daml start` or `./daml-start.sh` as we tell our users in the getting started guide?

File: ui/src/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  }
+  if (jsonApiProc) {
+    jsonApiProc.kill("SIGTERM");
+    console.log('Killed JSON API server');
+  }
+});
+
+test('create and look up user using ledger library', async () => {
+  const credentials = computeCredentials('Alice');
+  const ledger = new Ledger({token: credentials.token, httpBaseUrl: undefined, wsBaseUrl});
```
Comment (hurryabit) [inline ui/src/index.test.tsx]: The `--testURL` pass to `react-scripts test` makes sure this gets set up correctly.
```suggestion
  const ledger = new Ledger({token: credentials.token});
```

File: ui/src/index.test.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  await ledger.query(User);
```
Comment (hurryabit) [inline ui/src/index.test.tsx]: Why query if we don't do anything with the result?

File: ui/src/index.test.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    console.log('Killed sandbox');
+  }
+  if (jsonApiProc) {
+    jsonApiProc.kill("SIGTERM");
+    console.log('Killed JSON API server');
+  }
+});
+
+test('create and look up user using ledger library', async () => {
+  const credentials = computeCredentials('Alice');
```
Comment (hurryabit) [inline ui/src/index.test.tsx]:
```suggestion
  const {party, token} = computeCredentials('Alice');
```
This saves us from writing `credentials.` a few times further down.

File: ui/src/index.test.tsx (hunk 6 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: Hmm that's a mistake. Maybe I meant to test that there are no users at this point. Will fix

File: ui/src/index.test.tsx (hunk 4 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: No compelling reason. I thought I might need to configure arguments differently. But I suppose it makes sense to replicate the process that the user goes through, so I can change this.

File: ui/src/index.test.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  // Run daml process in create-daml-app root dir.
+  // The path should already include '.daml/bin' in the environment where this is run.
+  const opts: SpawnOptions = { cwd: '..', stdio: 'inherit' };
+
+  // Use `daml start` to start up the sandbox and json api server.
+  // This is what we recommend to our users (over running the two processes separately),
+  // so we replicate it in these tests.
+  const startArgs = [
+    'start',
+    '--open-browser=no',
```
Comment (cocreature) [inline ui/src/index.test.tsx]: This is redundant if you have `--start-navigator=n`.

File: ui/src/index.test.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    '--start-navigator=no',
+    '--sandbox-option=--wall-clock-time',
+    `--sandbox-option=--ledgerid=${LEDGER_ID}`,
+  ];
+  startProc = spawn('daml', startArgs, opts);
+
+  // We know that the processes are up and running once their ports become available.
+  await waitOn({resources: [`tcp:localhost:${SANDBOX_PORT}`, `tcp:localhost:${JSON_API_PORT}`]});
```
Comment (cocreature) [inline ui/src/index.test.tsx]: Can you wait on http resources? Might be a bit more robust to wait for the http json API to accept an http connection.

Comment (cocreature) [reply ui/src/index.test.tsx]: Also you don’t need to wait for sandbox. `daml start` only starts the json API once sandbox has finished running.

File: ui/src/index.test.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  const {party, token} = computeCredentials('Alice');
+  const ledger = new Ledger({token});
+  const users0 = await ledger.query(User);
+  expect(users0).toEqual([]);
+  const user: User = {username: party, friends: []};
+  const userContract1 = await ledger.create(User, user);
+  const userContract2 = await ledger.lookupByKey(User, party);
+  expect(userContract1).toEqual(userContract2);
+  const users = await ledger.query(User);
+  expect(users[0].contractId).toEqual(userContract1.contractId);
```
Comment (hurryabit) [inline ui/src/index.test.tsx]:
```suggestion
  expect(users[0]).toEqual(userContract1);
```

File: ui/src/index.test.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+  // We know that the processes are up and running once their ports become available.
+  await waitOn({resources: [`tcp:localhost:${SANDBOX_PORT}`, `tcp:localhost:${JSON_API_PORT}`]});
+}, 20_000);
+
+afterEach(() => {
+  // Shut down `daml start` process
+  // TODO: Test/fix this on windows
+  if (startProc) {
+    startProc.kill("SIGTERM");
```
Comment (cocreature) [inline ui/src/index.test.tsx]: Does `kill` wait for the process to terminate or just send the signal? You probably want to wait afterwards.

**Pull request merged**